retrieval/discovery/kubernetes_v2/pod.go

```diff
@@ -0,0 +1,142 @@
+package kubernetesv2
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+)
+
+// Pods discovers new pod targets.
+type Pod struct {
+	informer cache.SharedInformer
+	store    cache.Store
+	logger   log.Logger
+}
+
+// NewPods creates a new pod discovery.
+func NewPods(l log.Logger, pods cache.SharedInformer) *Pod {
+	return &Pod{
+		informer: pods,
+		store:    pods.GetStore(),
+		logger:   l,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (p *Pod) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	p.logger.Debugln("start")
+
+	// Send full initial set of pod targets.
+	var initial []*config.TargetGroup
+	add := func(tg *config.TargetGroup) {
+		p.logger.With("tg", fmt.Sprintf("%#v", tg)).Debugln("initial pod")
+		initial = append(initial, tg)
+	}
+	for _, o := range p.store.List() {
+		p.buildPod(o.(*apiv1.Pod), add)
+	}
+	select {
+	case <-ctx.Done():
+		return
+	case ch <- initial:
+	}
+
+	// Send target groups for pod updates.
+	send := func(tg *config.TargetGroup) {
+		p.logger.With("tg", fmt.Sprintf("%#v", tg)).Debugln("pod update")
+		select {
+		case <-ctx.Done():
+		case ch <- []*config.TargetGroup{tg}:
+		}
+	}
+	p.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(o interface{}) {
+			p.buildPod(o.(*apiv1.Pod), send)
+		},
+		DeleteFunc: func(o interface{}) {
+			send(&config.TargetGroup{Source: podSource(o.(*apiv1.Pod))})
+		},
+		UpdateFunc: func(_, o interface{}) {
+			p.buildPod(o.(*apiv1.Pod), send)
+		},
+	})
+
+	// Block until the target provider is explicitly canceled.
+	<-ctx.Done()
+}
+
+const (
+	podNameLabel                  = metaLabelPrefix + "pod_name"
+	podAddressLabel               = metaLabelPrefix + "pod_address"
```

Comment:
Generally, in other SDs we don't go as far as providing ip and port in their own labels in addition to the `__address__` label. Wondering in how far we should keep doing that here.